File: src/oca/java/src/org/opennebula/client/hook/Hook.java
/*******************************************************************************
 * Copyright 2002-2021, OpenNebula Project, OpenNebula Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ******************************************************************************/
package org.opennebula.client.hook;

import org.opennebula.client.Client;
import org.opennebula.client.OneResponse;
import org.opennebula.client.PoolElement;
import org.w3c.dom.Node;

/**
 * This class represents an OpenNebula Hook.
 * It also offers static XML-RPC call wrappers.
 */
public class Hook extends PoolElement
{

    private static final String METHOD_PREFIX = "hook.";
    private static final String ALLOCATE = METHOD_PREFIX + "allocate";
    private static final String INFO     = METHOD_PREFIX + "info";
    private static final String DELETE   = METHOD_PREFIX + "delete";
    private static final String UPDATE   = METHOD_PREFIX + "update";
    private static final String RENAME   = METHOD_PREFIX + "rename";
    private static final String LOCK     = METHOD_PREFIX + "lock";
    private static final String UNLOCK   = METHOD_PREFIX + "unlock";
    private static final String RETRY    = METHOD_PREFIX + "retry";

    /**
     * Creates a new Hook representation.
     * @param id The Hook id.
     * @param client XML-RPC Client.
     */
    public Hook(int id, Client client)
    {
        super(id, client);
    }

    /**
     * @see PoolElement
     */
    protected Hook(Node xmlElement, Client client)
    {
        super(xmlElement, client);
    }

    // =================================
    // Static XML-RPC methods
    // =================================


    /**
     * Allocates a new Hook in OpenNebula.
     *
     * @param client XML-RPC Client.
     * @param description A string containing the template of the Hook.
     * @return If successful the message contains the associated
     * id generated for this Hook.
     */
    public static OneResponse allocate(Client client, String description)
    {
        return client.call(ALLOCATE, description);
    }

    /**
     * Retrieves the information of the given Hook.
     *
     * @param client XML-RPC Client.
     * @param id The Hook id for the Hook to retrieve the information from
     * @return If successful the message contains the string
     * with the information returned by OpenNebula.
     */
    public static OneResponse info(Client client, int id)
    {
        return client.call(INFO, id, false);
    }

    /**
     * Deletes a Hook from OpenNebula.
     *
     * @param client XML-RPC Client.
     * @param id The Hook id of the target Hook we want to delete.
     * @return A encapsulated response.
     */
    public static OneResponse delete(Client client, int id)
    {
        return client.call(DELETE, id, false);
    }

    /**
     * Replaces the Hook contents.
     *
     * @param client XML-RPC Client.
     * @param id The Hook id of the target Hook we want to modify.
     * @param new_template New Hook contents.
     * @param append True to append new attributes instead of replace the whole Hook
     * @return If successful the message contains the Hook id.
     */
    public static OneResponse update(Client client, int id, String new_template,
        boolean append)
    {
        return client.call(UPDATE, id, new_template, append ? 1 : 0);
    }

    /**
     * Renames this Hook
     *
     * @param client XML-RPC Client.
     * @param id The Hook id of the target Hook.
     * @param name New name for the Hook.
     * @return If an error occurs the error message contains the reason.
     */
    public static OneResponse rename(Client client, int id, String name)
    {
        return client.call(RENAME, id, name);
    }

    /**
     * lock this Hook
     *
     * @param client XML-RPC Client.
     * @param id The hook id.
     * @param level Lock level.
     * @return If an error occurs the error message contains the reason.
     */
    public static OneResponse lock(Client client, int id, int level)
    {
        return client.call(LOCK, id, level);
    }

    /**
     * Unlock this Hook
     *
     * @param client XML-RPC Client.
     * @param id The Hook id.
     * @return If an error occurs the error message contains the reason.
     */
    public static OneResponse unlock(Client client, int id)
    {
        return client.call(UNLOCK, id);
    }

    /**
     * Retry this Hook
     *
     * @param client XML-RPC Client.
     * @param id The Hook id.
     * @param exec_id the hook execution id.
     * @return If an error occurs the error message contains the reason.
     */
    public static OneResponse retry(Client client, int id, int exec_id)
    {
        return client.call(RETRY, id, exec_id);
    }

    // =================================
    // Instanced object XML-RPC methods
    // =================================

    /**
     * Retrieves the information of the Hook.
     *
     * @return If successful the message contains the string
     * with the information returned by OpenNebula.
     */
    public OneResponse info()
    {
        OneResponse response = info(client, id);
        super.processInfo(response);
        return response;
    }

    /**
     * Deletes the Hook from OpenNebula.
     *
     * @return A encapsulated response.
     */
    public OneResponse delete()
    {
        return delete(client, id);
    }

    /**
     * Replaces the Hook contents.
     *
     * @param new_template New Hook contents.
     * @return If successful the message contains the Hook id.
     */
    public OneResponse update(String new_template)
    {
        return update(new_template, false);
    }

    /**
     * Replaces the Hook contents.
     *
     * @param new_template New Hook contents.
     * @param append True to append new attributes instead of replace the whole Hook
     * @return If successful the message contains the hook id.
     */
    public OneResponse update(String new_template, boolean append)
    {
        return update(client, id, new_template, append);
    }

    /**
     * Renames this Hook
     *
     * @param name New name for the Hook.
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse rename(String name)
    {
        return rename(client, id, name);
    }

    /**
     * Lock this Hook
     *
     * @param level Lock level.
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse lock(int level)
    {
        return lock(client, id, level);
    }

    /**
     * Unlock this Hook
     *
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse unlock()
    {
        return unlock(client, id);
    }

    /**
     * Retry this Hook
     *
     * @param exec_id the hook execution id.
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse retry(int exec_id)
    {
        return retry(client, id, exec_id);
    }
}

File: src/oca/ruby/opennebula/hook.rb
# -------------------------------------------------------------------------- #
# Copyright 2002-2024, OpenNebula Project, OpenNebula Systems                #
#                                                                            #
# Licensed under the Apache License, Version 2.0 (the "License"); you may    #
# not use this file except in compliance with the License. You may obtain    #
# a copy of the License at                                                   #
#                                                                            #
# http://www.apache.org/licenses/LICENSE-2.0                                 #
#                                                                            #
# Unless required by applicable law or agreed to in writing, software        #
# distributed under the License is distributed on an "AS IS" BASIS,          #
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   #
# See the License for the specific language governing permissions and        #
# limitations under the License.                                             #
#--------------------------------------------------------------------------- #

require 'opennebula/lockable_ext'
require 'opennebula/pool_element'

module OpenNebula

    # Class for representing a Hook object.
    class Hook < PoolElement

        #######################################################################
        # Constants and Class Methods
        #######################################################################
        HOOK_METHODS = {
            :allocate => 'hook.allocate',
            :delete   => 'hook.delete',
            :update   => 'hook.update',
            :rename   => 'hook.rename',
            :info     => 'hook.info',
            :lock     => 'hook.lock',
            :unlock   => 'hook.unlock',
            :retry    => 'hook.retry'
        }

        # Creates a Hook description with just its identifier
        # this method should be used to create plain Hook objects.
        # +id+ the id of the user
        #
        # Example:
        #   hook = Hook.new(Hook.build_xml(3),rpc_client)
        #
        def self.build_xml(pe_id = nil)
            if pe_id
                obj_xml = "<HOOK><ID>#{pe_id}</ID></HOOK>"
            else
                obj_xml = '<HOOK></HOOK>'
            end

            XMLElement.build_xml(obj_xml, 'HOOK')
        end

        # Class constructor
        def initialize(xml, client)
            LockableExt.make_lockable(self, HOOK_METHODS)

            super(xml, client)

            @client = client
        end

        #######################################################################
        # XML-RPC Methods for the Hook Object
        #######################################################################

        # Retrieves the information of the given Hook.
        def info
            return Error.new('ID not defined') unless @pe_id

            rc = @client.call(HOOK_METHODS[:info], @pe_id, false)

            if !OpenNebula.is_error?(rc)
                initialize_xml(rc, 'HOOK')
                rc = nil

                @pe_id = self['ID'].to_i if self['ID']
                @name  = self['NAME'] if self['NAME']
            end

            rc
        end

        alias :info! info

        # Allocates a new Hook in OpenNebula
        #
        # @param template [String] The contents of the Hook template.
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def allocate(template)
            super(HOOK_METHODS[:allocate], template)
        end

        # Deletes the Hook
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def delete
            call(HOOK_METHODS[:delete], @pe_id)
        end

        # Replaces the Hook contents
        #
        # @param new_template [String] New Hook contents
        # @param append [true, false] True to append new attributes instead of
        #   replace the whole template
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def update(new_template, append = false)
            super(HOOK_METHODS[:update], new_template, append ? 1 : 0)
        end

        # Renames this Hook
        #
        # @param name [String] New name for the Hook.
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def rename(name)
            call(HOOK_METHODS[:rename], @pe_id, name)
        end

        # Retry a previous execution of the hook.
        #
        # @param exec_id [int] Hook execution id.
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def retry(exec_id)
            call(HOOK_METHODS[:retry], @pe_id, exec_id)
        end

        #######################################################################
        # Helpers to get Hook information
        #######################################################################

        # Returns the group identifier
        # [return] _Integer_ the element's group ID
        def gid
            self['GID'].to_i
        end

        def owner_id
            self['UID'].to_i
        end

    end

end
